design/apistyle-no2.md

```diff
@@ -0,0 +1,496 @@
+
+This cdk8s Chart would synth the content for the following XRDs and backing `Compositions`:
+
+* `CompositeCluster.aws.platform.acme.io`
+* `EKS.aws.platform.acme.io`
+* `Services.aws.platform.acme.io`
+
+```ts
+import { Chart } from 'cdk8s';
+import { Construct } from 'constructs';
+import { PropType, MetaUI } from 'crossplane-cdk8s';
+import * as crossplane from 'crossplane-cdk8s';
+import * as iam from '../imports/identity.aws.crossplane.io';
+import * as eks from '../imports/eks.aws.crossplane.io';
+import * as helm from '../imports/helm.crossplane.io';
+
+const PATH_METADATA_UID = 'metadata.uid';
+const PATH_METADATA_NAME = 'metadata.name';
+const PATH_METADATA_LABELS = 'metadata.labels';
+const PATH_METADATA_ANNOTATIONS = 'metadata.annotations'
+const PATH_EXTERNAL_NAME = 'metadata.annotations[crossplane.io/external-name]';
+const PATH_CONN_SECRET_NAME = 'spec.writeConnectionSecretToRef.name';
+const PATH_CONN_SECRET_NAMESPACE = 'spec.writeConnectionSecretToRef.namespace';
+
+enum NodeSize {
+  SMALL = 'small',
+  MEDIUM = 'medium',
+  LARGE = 'large',
+}
+
+export class ClusterChart extends Chart {
+
+  private id: StringProp = stringProp({
+    path: 'spec.id',
+    description: 'Cluster ID that other objects will use to refer to this cluster',
+    requireAllInPath: true,
+  });
+
+  private networkRef: StringProp = stringProp({
+    path: 'spec.parameters.networkRef.id',
+    description: 'A reference to the Network object that this Cluster should be connected to.',
+    requireAllInPath: true,
+  });
+
+  private nodeCount: IntegerProp = integerProp({
+    path: 'spec.parameters.nodes.count',
+    description: 'Desired node count, from 1 to 100.',
+    requireAllInPath: true,
+  });
+
+  private nodeSize: EnumProp = enumProp({
+    path: 'spec.parameters.nodes.size',
+    description: 'Size of node.',
+    requireAllInPath: true,
+    enum: [
+      NodeSize.SMALL,
+      NodeSize.MEDIUM,
+      NodeSize.LARGE,
+    ]
+  });
+
+  private promVersion: StringProp = stringProp({
+    path: 'spec.parameters.services.operators.prometheus.version',
+    description: 'The version of kube-prometheus-stack chart to install.',
+    requireAllInPath: true,
+  });
+
+  private connectionSecret: ConnectionSecretProps = {
+    defaultNamespace: 'crossplane-system',
+    keys: [`kubeconfig`]
+  };
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    // Cluster XRD
+    const xrdCluster = new crossplane.CompositeResourceDefinition(this, 'cluster-xrd', {
+      name: 'compositeclusters.aws.platform.acme.io',
+      group: 'aws.platform.acme.io',
+      claimKind: {
+        name: 'Cluster',
+        plural: 'clusters',
+      },
+      kind: {
+        name: 'CompositeCluster',
+        plural: 'compositeclusters'
+      },
+      connectionSecret: this.connectionSecret,
+      versions: [
+        {
+          name: 'v1alpha1',
+          served: true,
+          referenceable: true,
+          schemaSpec: [
```

Comment:
I am wondering if we can leverage some existing json schema tools here but that could be in the future.

For example, we could generate the schema from typescript interfaces.

Reply:
good idea. I wanted to keep the XRD definition succinct so you could see the entire definition in a single screen and having inline schema definitions was cumbersome and didn't cleanly support prop auto mapping from parent XRD to composed XRD with the same schema props (or with path munging). If we can generate schemas from interfaces defined in the same file (e.g. we need to compose XRDs in compositions as well) and support prop auto mapping with path munging I'd be all for it.